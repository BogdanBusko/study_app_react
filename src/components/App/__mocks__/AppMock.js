export const userInfo = {
  data: {
    data: {
      user: {
        id: 1,
        type: 'user',
        attributes: {
          first_name: 'John',
          last_name: 'Doe',
          full_name: 'John Doe'
        }
      }
    }
  }
}
