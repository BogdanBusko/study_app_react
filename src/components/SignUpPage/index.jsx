import { useForm } from 'react-hook-form';
import { useDispatch } from 'react-redux';
import {
  Grid,
  TextField,
  Button
} from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';

import signUpUser from '../../redux/actions/signUpUser';

const useStyles = makeStyles({
  root: {
    paddingTop: '20px',
  },
})

const SignUpPage = () => {
  const { register, handleSubmit } = useForm();
  const dispatch = useDispatch();
  const classes = useStyles();

  const onSubmit = (data) => {
    dispatch(signUpUser(data));
  }

  return(
    <Grid
      container
      alignItems="center"
      justify="center"
      className={classes.root}
    >
      <form onSubmit={handleSubmit(onSubmit)}>
        <Grid container direction="column" spacing={2}>
          <TextField
            name="email"
            ref={register}
            variant="outlined"
            label="Email"
            margin="normal"
          />
          <TextField
            name="password"
            ref={register}
            variant="outlined"
            label="Password"
            margin="normal"
          />
          <Button variant="contained">
            Sign Up
          </Button>
        </Grid>
      </form>
    </Grid>
  );
}

export default SignUpPage;
