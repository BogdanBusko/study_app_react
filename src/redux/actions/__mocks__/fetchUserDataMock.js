export const userInfo = {
  data: {
    data: {
      id: 1,
      type: 'user',
      attributes: {
        first_name: 'John',
        last_name: 'Doe',
        full_name: 'John Doe'
      }
    }
  }
}
