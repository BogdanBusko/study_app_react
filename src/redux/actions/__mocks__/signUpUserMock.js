export const registrationInfo = {
  data: {
    data: {
      id: '1',
      type: 'registration',
      attributes: {
        token: 'eyJhbGciOiJIUzI1NiJ9.eyJkYXRhIjp7InVzZXJfaWQiOjF9fQ.2MsjQvcIt3wDxROaQN21aMFNgtdYZCYCpJrlNwmYkcc',
        user: {
          data: {
            id: 1,
            type: 'user',
            attributes: {
              first_name: 'John',
              last_name: 'Doe',
              full_name: 'John Doe'
            }
          }
        }
      }
    }
  }
}
