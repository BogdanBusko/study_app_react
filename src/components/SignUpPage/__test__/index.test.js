import React from 'react';
import { fireEvent, waitFor, act } from '@testing-library/react';

import fetchData from '../../../config/axios';
import { customRender } from '../../../test_helpers/helpers';
import SignUpPage from '../index';
import * as mock from '../__mocks__/SignUpPageMock';

jest.mock('../../../config/axios');

it("renders correctly", () => {
  const { asFragment } = customRender(<SignUpPage />);

  expect(asFragment(<SignUpPage />)).toMatchSnapshot();
});

describe("with invalid data", () => {
  it("does not submit form when user does not fill all requeired fileds", async () => {
    const { queryByText, getByTestId } = customRender(<SignUpPage />);

    fireEvent.click(getByTestId('signup-button'));

    await waitFor(() => expect(queryByText(/First name can't be blank/i)).toBeTruthy());
    await waitFor(() => expect(queryByText(/Last name can't be blank/i)).toBeTruthy());
    await waitFor(() => expect(queryByText(/Email can't be blank/i)).toBeTruthy());
    await waitFor(() => expect(queryByText(/Password can't be blank/i)).toBeTruthy());
    await waitFor(() => expect(queryByText(/Password confirmation can't be blank/i)).toBeTruthy());
  });

  it("does not submit form if password is to short", async () => {
    const { queryByText, getByTestId } = customRender(<SignUpPage />);
    const passwordInput = getByTestId('password').querySelector('input');

    fireEvent.change(passwordInput, { target: { value: '456' } });

    await act(async () => fireEvent.click(getByTestId('signup-button')));

    await waitFor(() => expect(queryByText(/Password must contain at least 6 characters/i)).toBeTruthy());
  });

  it("does not submit form if password and password confirmation does not match", async () => {
    const { queryByText, getByTestId } = customRender(<SignUpPage />);
    const passwordInput = getByTestId('password').querySelector('input');
    const passwordConfirmationInput = getByTestId('password_confirmation').querySelector('input');

    fireEvent.change(passwordInput, { target: { value: '123456' } });
    fireEvent.change(passwordConfirmationInput, { target: { value: '112345' } });

    await act(async () => fireEvent.click(getByTestId('signup-button')));

    await waitFor(() => expect(queryByText(/The password does not match/i)).toBeTruthy());
  });
});

describe("with valid data", () => {
  it("successfully sign up", async () => {
    const { getByTestId } = customRender(<SignUpPage />);
    const firstName = getByTestId('first_name').querySelector('input');
    const lastName = getByTestId('last_name').querySelector('input');
    const email = getByTestId('email').querySelector('input');
    const passwordInput = getByTestId('password').querySelector('input');
    const passwordConfirmationInput = getByTestId('password_confirmation').querySelector('input');

    fetchData.post.mockResolvedValue(mock.successSignUp);

    await act(async () => {
      fireEvent.change(firstName, { target: { value: 'John' } });
      fireEvent.change(lastName, { target: { value: 'Doe' } });
      fireEvent.change(email, { target: { value: 'email@gmail.com' } });
      fireEvent.change(passwordInput, { target: { value: '123456' } });
      fireEvent.change(passwordConfirmationInput, { target: { value: '123456' } });
    });

    setTimeout(async () => {
      await act(async () => fireEvent.click(getByTestId('signup-button')));
    }, 1);

    await waitFor(() =>
      expect(localStorage.getItem('token')).toEqual(mock.successSignUp.data.data.attributes.token)
    );
  });
});
