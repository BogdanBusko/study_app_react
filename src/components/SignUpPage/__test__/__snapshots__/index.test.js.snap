// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root makeStyles-root-1 MuiGrid-container MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
  >
    <form>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-direction-xs-column"
      >
        <div
          class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal"
          data-testid="first_name"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
            data-shrink="false"
          >
            First name
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input"
              name="first_name"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root-2 MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="PrivateNotchedOutline-legendLabelled-4"
              >
                <span>
                  First name
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal"
          data-testid="last_name"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
            data-shrink="false"
          >
            Last name
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input"
              name="last_name"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root-2 MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="PrivateNotchedOutline-legendLabelled-4"
              >
                <span>
                  Last name
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal"
          data-testid="email"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
            data-shrink="false"
          >
            Email
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input"
              name="email"
              type="email"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root-2 MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="PrivateNotchedOutline-legendLabelled-4"
              >
                <span>
                  Email
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal"
          data-testid="password"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
            data-shrink="false"
          >
            Password
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input"
              name="password"
              type="password"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root-2 MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="PrivateNotchedOutline-legendLabelled-4"
              >
                <span>
                  Password
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal"
          data-testid="password_confirmation"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
            data-shrink="false"
          >
            Password confirmation
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input"
              name="password_confirmation"
              type="password"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root-2 MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="PrivateNotchedOutline-legendLabelled-4"
              >
                <span>
                  Password confirmation
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained"
          data-testid="signup-button"
          tabindex="0"
          type="submit"
        >
          <span
            class="MuiButton-label"
          >
            Sign Up
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
